lib/Rabbit.js

```diff
@@ -0,0 +1,253 @@
+     * },
+     */
+};
+
+function getCompleteConfig(config) {
+    const completeConfig = Object.assign({}, defaultConfig, config);
+    return completeConfig;
+}
+
+function getConnectionString(config) {
+    const { secure, user, password, host, port, connectionString } = config.connection;
+    if (connectionString) {
+        return connectionString;
+    } else {
+        const protocol = (typeof secure === "boolean" && secure) ? "amqps" : "amqp";
+        const protocolBasedPort = protocol === "amqps" ? 5671 : 5672;
+        if (user && password) {
+            return `${protocol}://${user}:${password}@${host}:${port}`;
+        }
+        return `${protocol}://${host}:${port}`;
+    }
+}
+
+async function sleep(delay) {
+    return new Promise((resolve, reject) => {
+        setTimeout(resolve, delay);
+    });
+}
+
+class Rabbit {
+
+    constructor(config = {}, logger = console) {
+        this._publishedQueueList = []; // queue name that has been published (to avoid redundant publication)
+        this._connection = null;
+        this._config = getCompleteConfig(config);
+        this._logger = logger;
+        this._connectionString = getConnectionString(this._config);
+        this._run();
+    }
+
+    async publish(payload) {
+        if (this._connection) {
+            this._config.pubPayload = payload;
+            this._publishedQueueList = [];
+            await this._pubQueue(this._connection);
+            await this._pub(this._connection);
+        } else {
+            await sleep(1000);
+            await this.publish(payload);
+        }
+    }
+
+    async _run() {
+        try {
+            // create connection
+            const connection = await amqp.connect(this._connectionString);
+            this._logger.info("Connected to rmq.");
+            connection.on("error", async (error) => {
+                this._connection = null;
+                this._logger.error("Connection error : ", error);
+                await sleep(1000);
+                await this._run();
+            });
+            connection.on("close", async (error) => {
+                this._connection = null;
+                this._logger.error("Connection close : ", error);
+                await sleep(1000);
+                await this._run();
+            });
+            this._connection = connection;
+            // process pubSub handler
+            await this._subQueue(this._connection);
+            await this._sub(this._connection);
+            await this._pubQueue(connection);
+            await this._pub(connection);
+            // add to channels
+        } catch (error) {
+            this._connection = null;
+            this._logger.warn("Failed to connect to rmq.", error);
+            await sleep(1000);
+            this._logger.info("Attempting to reconnect to rmq.");
+            await this._run();
+        }
+    }
+
+    async _pub(connection) {
+        if (!this._config.pub || !this._connection) {
```

Comment:
if we want this method to be more "isolated", let's use `connection` from params instead of `this._connection` from instance

Reply:
suggestion accepted